community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/ForkedProcessorStep.java

```diff
@@ -266,25 +281,19 @@ public void run()
         {
             try
             {
-                while ( !isCompleted() )
+                while ( !isCompleted() && !end )
                 {
                     Unit candidate = current.next;
                     if ( candidate != null )
                     {
+                        // There's work to do
                         if ( id < candidate.processors )
                         {
-                            // There's work to do
                             long time = nanoTime();
                             forkedProcess( id, candidate.processors, candidate.batch );
                             candidate.processorDone( nanoTime() - time );
                         }
-                        else
-                        {
-                            // The id of this processor is greater than that of the next unit's expected max.
-                            // This means that the number of assigned processors to this step has decreased
-                            // and that this processor have reached the end of its life.
-                            break;
-                        }
```

Comment:
This is actually by design. There was a whole pile of races involved with actually removing the processor threads, since new threads could be created holding the same ID. We will reuse the threads if the processor count first decreases and then increases again, so we considered it to be an acceptable trade-off.